File: backend/mcp-code-intelligence-nodejs/src/memory/embedding/ollama-provider.ts
/**
 * Ollama-backed embedding provider — wraps HTTP calls to Ollama API.
 * Uses /api/embeddings endpoint for text-to-vector conversion.
 */

import { EmbeddingProvider } from './provider.js';

/** Minimal Ollama client interface for embedding. */
export interface OllamaClient {
  getEmbedding(text: string): Promise<number[] | null>;
  isAvailable(): boolean;
}

export class OllamaEmbeddingProvider implements EmbeddingProvider {
  private readonly client: OllamaClient;
  private readonly model: string;

  constructor(client: OllamaClient, model: string) {
    this.client = client;
    this.model = model;
  }

  get modelName(): string {
    return this.model;
  }

  get dimensions(): number {
    // nomic-embed-text default dimension
    return 768;
  }

  async embed(text: string): Promise<number[] | null> {
    return this.client.getEmbedding(text);
  }

  async embedBatch(texts: string[]): Promise<Array<number[] | null>> {
    const results: Array<number[] | null> = [];
    for (const text of texts) {
      results.push(await this.embed(text));
    }
    return results;
  }

  isAvailable(): boolean {
    return this.client.isAvailable();
  }

  close(): void {
    // No resources to release for HTTP client
  }
}

function log(msg: string): void {
  process.stderr.write(`[ollama-embed] ${msg}\n`);
}

File: backend/mcp-code-intelligence-python/src/mcp_code_intel/memory/embedding/ollama_provider.py
"""Ollama-backed embedding provider — wraps existing OllamaClient."""

import sys

from .provider import EmbeddingProvider
from ...ollama import OllamaClient


class OllamaEmbeddingProvider(EmbeddingProvider):
    """Delegates embedding to remote Ollama server."""

    def __init__(self, client: OllamaClient, model: str) -> None:
        self._client = client
        self._model = model

    @property
    def model_name(self) -> str:
        return self._model

    @property
    def dimensions(self) -> int:
        # nomic-embed-text default dimension
        return 768

    def embed(self, text: str) -> list[float] | None:
        """Get embedding from Ollama API."""
        return self._client.get_embedding(text)

    def is_available(self) -> bool:
        """Check if Ollama server is reachable."""
        return self._client.is_available()

    def close(self) -> None:
        pass


def _log(msg: str) -> None:
    print(f"[ollama-embed] {msg}", file=sys.stderr, flush=True)
